File: unicallPY/src/unicall/client/spinup.py
import os
import socket
import struct
import subprocess

# Custom function and return handling (similar to `Function` and `Magic` in Node.js)
class Function:
    def __init__(self, types, ret, id, name):
        self.types = types
        self.ret = ret
        self.id = id
        self.name = name

class Library:
    def __init__(self):
        self.functions = None
        self.waitlist = []  # To store return data callbacks

# Setup the UNIX socket server
def setup_socket(library, socket_name):
    # Remove the socket file if it already exists
    if os.path.exists(socket_name):
        os.remove(socket_name)
    
    # Create a UNIX socket
    server_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    server_socket.bind(socket_name)
    server_socket.listen()

    # Run the server
    run_server(library, socket_name)

    # Accept and handle connections
    while True:
        conn, _ = server_socket.accept()
        with conn:
            while True:
                header = conn.recv(5)
                if not header:
                    break
                
                # Parse header to check message type and length
                msg_type = header[0]
                length = (header[1] << 24) + (header[2] << 16) + (header[3] << 8) + header[4]
                data = conn.recv(length)

                # Handle function definitions or return data based on msg_type
                if msg_type == 0xF2:
                    get_functions(library, data, length)
                elif msg_type == 0xF1:
                    process_return(library, data)

# Launch a subprocess for the child server
def run_server(library, socket_name):
    # Assuming PY is a predefined constant for Python
    subprocess.Popen(["python", library.filename, f"socket={socket_name}"])

def get_functions(library, data, length):
    library.functions = []
    sidx = 0
    while sidx < length:
        sidx = get_function(library.functions, data, sidx)

def get_function(function_list, data, s):
    id = (data[s] << 8) + data[s+1]
    s += 2
    slen = (data[s] << 8) + data[s+1]
    s += 2
    name = data[s:s+slen].decode('utf-8')
    s += slen
    ret = data[s]
    s += 1
    arglen = (data[s] << 8) + data[s+1]
    s += 2
    types = list(data[s:s+arglen])
    s += arglen
    function_list.append(Function(types, ret, id, name))
    return s

def process_return(library, data):
    s = 0
    retid = (data[s] << 8) + data[s+1]
    s += 2
    value = decode(data[s:])  # Assuming a decode function that handles data

    idx = -1
    for i, (rid, future) in enumerate(library.waitlist):
        if rid == retid:
            idx = i
            future.resolve(value)
            break

    if idx == -1:
        raise ValueError(f"Return ID {retid} not found in waitlist")
    else:
        library.waitlist.pop(idx)

# Example decode function (to be customized based on data structure)
def decode(data):
    # This function should interpret `data` based on expected encoding
    # Here we assume `data` is a UTF-8 encoded string for simplicity
    return data.decode('utf-8')

# Main function to run the server setup
def main():
    library = Library()
    library.filename = "script.py"  # The Python script to run
    socket_name = "/tmp/my_socket"

    try:
        setup_socket(library, socket_name)
    finally:
        if os.path.exists(socket_name):
            os.remove(socket_name)

if __name__ == "__main__":
    main()

File: unicallJS/client/spinup.mjs
import * as net from "net"
import * as fs from "fs"
import { exec } from "child_process"
import { Function, PY } from "../classes.mjs"
import { strict as assert } from "assert"
import { decode } from "../coding.mjs"
import { Magic } from "../magic.mjs"

export async function setup_socket(library, socket_name) {
  const function_def_finished = new Promise((res, _) => {

    if (fs.existsSync(socket_name)) fs.rmSync(socket_name)
    library.server = net.createServer((socket) => {
      library.socket = socket
      socket.on("readable", () => {
        // first we have to get the functions
        if (library.functions === undefined) {
          const read = socket.read(5)
          if (read == null) return
          const head = new Uint8Array(read)

          assert.equal(head[0], Magic.FDEF)
          const length = (head[1] << 24) + (head[2] << 16) + (head[3] << 8) + head[4]
          const data = new Uint8Array(socket.read(length))

          get_functions(library, data, length)
          return res()
        }

        // otherwise it's a return
        const read = socket.read(5)
        if (read === null) return
        const head = new Uint8Array(read)
        assert.equal(head[0], Magic.RET)
        const length = head[1] << 24 | head[2] << 16 | head[3] << 8 | head[4]
        const data = new Uint8Array(socket.read(length))
        process_return(library, data, length)
      })
    })

    //first we listen
    library.server.listen(socket_name)
    //then we run the server
    run_server(library.filetype, library.filename, socket_name)
  })
  
  await function_def_finished
}

function run_server(type, file, socket_name) {
  switch (type) {
    case PY:
      exec(`python3 ${file} socket=${socket_name} > out 2>&1`)
      break;
    default:
      throw new Error("bad language")
  }
}

function get_functions(library, data, length) {

  library.functions = []
  let sidx = 0;
  while (sidx < length) {
    sidx = get_function(library.functions, data, sidx)
  }
}


function get_function(list, data, s) {
  const id = (data[s++] << 8) + data[s++]
  let slen = (data[s++] << 8) + data[s++] + s
  let name = ""
  for (; s<slen; s++) {
    name += String.fromCharCode(data[s])
  }
  const ret = data[s++]
  let arglen = (data[s++] << 8) + data[s++] + s
  const types = []
  for (; s < arglen; s++) {
    types.push(data[s])
  }
  
  list.push(new Function(types, ret, id, name))
  return s
  
}

function process_return(library, data) {
  let s=0
  const retid = (data[s++] << 8) + data[s++]
  console.log("RECIEVED ENCODED DATA: ", data)
  const value = decode(data, 2)
  console.log("RET VALUE", value)

  let idx;
  library.waitlist.forEach((a, i) => {
    if (a[0] != retid) return
    idx = i
    a[1](value)
  })

  if (idx === undefined)
    throw new Error(`return id ${retid} not found in waitlist`);
  library.waitlist.splice(idx, 1)
}










